visualization/app/codeCharta/ui/codeMap/threeViewer/threeSceneService.ts

```diff
@@ -76,47 +82,39 @@ export class ThreeSceneService implements CodeMapPreRenderServiceSubscriber {
 		this.getMapMesh().highlightBuilding(this.highlighted, this.selected, state, this.constantHighlight)
 	}
 
-	private updateCorrectMaterial(materials: Material[], nodes: CodeMapBuilding[], select: boolean) {
-		// TODO This loops introduce performance issues
-		//  refactor it please
-		return
-		const allNodes = this.mapMesh.getNodes()
-
-		for (const node of allNodes) {
-			for (const material of materials) {
-				for (const hoveredNode of nodes) {
-					if (hoveredNode !== null) {
-						if (select) {
-							if (material.userData.id === hoveredNode.node.id) {
-								material["color"].setHex(ColorConverter.convertHexToNumber(hoveredNode.color))
-							}
-						} else if (material.userData.id === hoveredNode.node.id) {
-							material["color"].setHex(ColorConverter.convertHexToNumber("#FFFFFF"))
-						} else if (material.userData.id === node.id) {
-							material["color"].setHex(ColorConverter.convertHexToNumber("#7A7777"))
-						}
-					}
-				}
-			}
-		}
+	private selectMaterial(materials: Material[]) {
+		const selectedMaterial = materials.find(({ userData }) => userData.id === this.selected.node.id)
+		selectedMaterial?.["color"].setHex(this.numberOrangeColor)
 	}
 
 	private resetMaterial(materials: Material[]) {
-		// TODO activate the function when the performance issue in updateCorrectMaterial() has been fixed.
-		return
-		const allNodes = this.mapMesh.getNodes()
-		for (const node of allNodes) {
-			for (const material of materials) {
-				if (material.userData.id === node.id) {
-					material["color"].setHex(ColorConverter.convertHexToNumber("#FFFFFF"))
-				}
+		const selectedID = this.selected ? this.selected.node.id : -1
+		for (const material of materials) {
+			const materialNodeId = material.userData.id
+			if (this.allNodesIds.has(materialNodeId) && materialNodeId !== selectedID) {
```

Comment:
Thinking about this again: is it possible to have material ids that do not exist in `allNodesIds`?

Reply:
Relevant one?, 
No I specifically store only correct ones in another code part